src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,341 @@
+#include <app/CommandHandlerInterface.h>
+#include <protocols/interaction_model/StatusCode.h>
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Actions {
+
+static constexpr size_t kActionNameMaxSize       = 128u;
+static constexpr size_t kEndpointListNameMaxSize = 128u;
+static constexpr size_t kEndpointListMaxSize     = 256u;
+
+class Delegate;
+
+struct ActionStructStorage : public Structs::ActionStruct::Type
+{
+    ActionStructStorage() : mActionName(mBuffer, sizeof(mBuffer)){};
+
+    ActionStructStorage(uint16_t action, CharSpan actionName, ActionTypeEnum actionType, uint16_t epListID,
+                        BitMask<CommandBits> commands, ActionStateEnum actionState) :
+        mActionName(mBuffer, sizeof(mBuffer))
+    {
+        Set(action, actionName, actionType, epListID, commands, actionState);
+    }
+
+    ActionStructStorage(const ActionStructStorage & action) : mActionName(mBuffer, sizeof(mBuffer)) { *this = action; }
+
+    ActionStructStorage & operator=(const ActionStructStorage & action)
+    {
+        Set(action.actionID, action.name, action.type, action.endpointListID, action.supportedCommands, action.state);
+        return *this;
+    }
+
+    void Set(uint16_t action, CharSpan actionName, ActionTypeEnum actionType, uint16_t epListID, BitMask<CommandBits> commands,
+             ActionStateEnum actionState)
+    {
+        actionID          = action;
+        type              = actionType;
+        endpointListID    = epListID;
+        supportedCommands = commands;
+        state             = actionState;
+        CopyCharSpanToMutableCharSpanWithTruncation(actionName, mActionName);
+        name = mActionName;
+    }
+
+private:
+    char mBuffer[kActionNameMaxSize];
+    MutableCharSpan mActionName;
+};
+
+struct EndpointListStorage : public Structs::EndpointListStruct::Type
+{
+    EndpointListStorage() : mEpListName(mBuffer, sizeof(mBuffer)){};
+
+    EndpointListStorage(uint16_t epListId, CharSpan epListName, EndpointListTypeEnum epListType,
+                        DataModel::List<const EndpointId> endpointList) :
+        mEpListName(mBuffer, sizeof(mBuffer))
+    {
+        Set(epListId, epListName, epListType, endpointList);
+    }
+
+    EndpointListStorage(const EndpointListStorage & epList) : mEpListName(mBuffer, sizeof(mBuffer)) { *this = epList; }
+
+    EndpointListStorage & operator=(const EndpointListStorage & epList)
+    {
+        Set(epList.endpointListID, epList.name, epList.type, epList.endpoints);
+        return *this;
+    }
+
+    void Set(uint16_t epListId, CharSpan epListName, EndpointListTypeEnum epListType,
+             DataModel::List<const EndpointId> endpointList)
+    {
+        endpointListID = epListId;
+        type           = epListType;
+
+        for (uint8_t index = 0; index < std::min(endpointList.size(), kEndpointListMaxSize); index++)
+        {
+            mEpList.push_back(endpointList[index]);
```

Comment:
This has terrible realloc behavior.  If we know the size, why are we not reserving it?

Also, this will crash on allocation failure the way this is written right now, as far as I can tell.  Is that intended?

Reply:
Removed vector usage